File: Intermediate/algorithms/mergesort.js
var unsortedArr = [340, 1, 3, 3, 76, 23, 4, 12, 122, 7642, 646];
function merge(leftArr, rightArr) {
  var sortedArr = [];
  while (leftArr.length && rightArr.length) {
    if (leftArr[0] <= rightArr[0]) {
      sortedArr.push(leftArr[0]);
      leftArr = leftArr.slice(1);
    } else {
      sortedArr.push(rightArr[0]);
      rightArr = rightArr.slice(1);
    }
  }
  while (leftArr.length) sortedArr.push(leftArr.shift());
  while (rightArr.length) sortedArr.push(rightArr.shift());
  return sortedArr;
}
function mergesort(arr) {
  if (arr.length < 2) {
    return arr;
  } else {
    var midpoint = parseInt(arr.length / 2);
    var leftArr = arr.slice(0, midpoint);
    var rightArr = arr.slice(midpoint, arr.length);
    return merge(mergesort(leftArr), mergesort(rightArr));
  }
}
console.log("This should be the sorted array!");
console.log(mergesort(unsortedArr));

File: Intermediate/algorithms/C++/merge_sort.cpp
#include<bits/stdc++.h>

using namespace std;

void merge(int arr[],int p,int q,int r)
{
	int n1 = q - p + 1;
	int n2 = r - q;
	int L[n1+1],R[n2+1];
	for(int i=0;i<n1;i++)
	{
		L[i] = arr[p+i];
	}
	for(int i=0;i<n2;i++)
	{
		R[i] = arr[q+i+1];
	}
	L[n1] = INT_MAX;
	R[n2] = INT_MAX;

	int i=0,j=0;
	for(int k=p;k<=r;k++)
	{
		if(L[i] <= R[j])
		{
			arr[k] = L[i];
			i++;
		}
		else
		{
			arr[k] = R[j];
			j++;
		}
	}
}
void merge_sort(int arr[],int p,int r)
{
	if(p<r)
	{
		int q = (p+r)/2;
		merge_sort(arr,p,q);
		merge_sort(arr,q+1,r);
		merge(arr,p,q,r);
	}
}
int main()
{
	int arr[] = {56,9,45,38,97,13,4,67,109,22};
	for(auto x : arr)
		cout<<x<<" ";
	cout<<endl;
	merge_sort(arr,0,9);
	for(auto x : arr)
	{
		cout<<x<<" ";
	}
	cout<<endl;
}